## Problem 1

> If we list all the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6 and 9. The sum of these multiples is 23.
>
> Find the sum of all the multiples of 3 or 5 below 1000.


In [ ]:
# This repeats some code, better to do it the other way
if number % 3 == 0:
    my_multiples.append(number)
elif number % 5 == 0:
    my_multiples.append(number)

In [11]:
my_multiples = []
for number in range(1, 1000):
    if (number % 3 == 0) or (number % 5 == 0):
        my_multiples.append(number)

print(my_multiples)

[3, 5, 6, 9, 10, 12, 15, 18, 20, 21, 24, 25, 27, 30, 33, 35, 36, 39, 40, 42, 45, 48, 50, 51, 54, 55, 57, 60, 63, 65, 66, 69, 70, 72, 75, 78, 80, 81, 84, 85, 87, 90, 93, 95, 96, 99, 100, 102, 105, 108, 110, 111, 114, 115, 117, 120, 123, 125, 126, 129, 130, 132, 135, 138, 140, 141, 144, 145, 147, 150, 153, 155, 156, 159, 160, 162, 165, 168, 170, 171, 174, 175, 177, 180, 183, 185, 186, 189, 190, 192, 195, 198, 200, 201, 204, 205, 207, 210, 213, 215, 216, 219, 220, 222, 225, 228, 230, 231, 234, 235, 237, 240, 243, 245, 246, 249, 250, 252, 255, 258, 260, 261, 264, 265, 267, 270, 273, 275, 276, 279, 280, 282, 285, 288, 290, 291, 294, 295, 297, 300, 303, 305, 306, 309, 310, 312, 315, 318, 320, 321, 324, 325, 327, 330, 333, 335, 336, 339, 340, 342, 345, 348, 350, 351, 354, 355, 357, 360, 363, 365, 366, 369, 370, 372, 375, 378, 380, 381, 384, 385, 387, 390, 393, 395, 396, 399, 400, 402, 405, 408, 410, 411, 414, 415, 417, 420, 423, 425, 426, 429, 430, 432, 435, 438, 440, 441, 444, 445, 447, 450,

In [12]:
# This block of code sums a sequence of numbers
result = 0
for number in my_multiples:
    result = result + number
print(result)

233168


In [14]:
# Same as above
sum(my_multiples)

233168

In [15]:
# Solution to problem 1
my_multiples = []
for number in range(1, 1000):
    if (number % 3 == 0) or (number % 5 == 0):
        my_multiples.append(number)

print(sum(my_multiples))

233168


## Problem 6

> The sum of the squares of the first ten natural numbers is,
> $$1^2 + 2^2 + \dots = 385$$
> The square of the sum of the first ten natural numbers is,
>
> Hence the difference between the sum of the squares of the first ten natural numbers and the square of the sum is
> 
>
> Find the difference between the sum of the squares of the first one hundred natural numbers and the square of the sum.

In [24]:
# Square of the sum
square_of_sum = sum(range(101)) ** 2

squares = []  # Empty list
for number in range(101):
    next_square = number ** 2
    squares.append(next_square)

sum_of_squares = sum(squares)

print(square_of_sum)
print(sum_of_squares)
print(square_of_sum - sum_of_squares)

25502500
338350
25164150


## Problem 2

> Each new term in the Fibonacci sequence is generated by adding the previous two terms. By starting with 1 and 2, the first 10 terms will be:
>
> 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, ...
>
> By considering the terms in the Fibonacci sequence whose values do not exceed four million, find the sum of the even-valued terms.


In [31]:
limit = 90

x = 1
y = 2
print(x, y)

z = x + y
print(z)

z = y + z
print(z)

z = y + z
print(z)

1 2
3
5
7


In [45]:
limit = 90
fibonacci = [1, 2]
print(limit)

while fibonacci[-1] + fibonacci[-2] < limit:
    fibonacci.append(fibonacci[-1] + fibonacci[-2])

fibonacci

90


[1, 2, 3, 5, 8, 13, 21, 34, 55, 89]

In [46]:
# Even terms of the Fibonacci sequence
fibonacci_even = []
for number in fibonacci:
    if number % 2 == 0:
        fibonacci_even.append(number)

sum(fibonacci_even)

44

In [47]:
x, y = 1, 2
print(x, y)

1 2


In [60]:
x, y = y, x + y
print(x, y)

610 987


## Problem 4

> A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.
>
> Find the largest palindrome made from the product of two 3-digit numbers.

In [63]:
max_element = 0
for number in [10, 1, 99, 8, -1, 100]:
    if number > max_element:
        max_element = number
max_element

100

In [62]:
max([10, 1, 99, 8, -1, 100])

100

In [70]:
"Juan Luis"[::-1]

'siuL nauJ'

In [72]:
number = 9009
str(number)[::-1] == str(number)

True

In [80]:
palindromes = []

for number_a in range(100, 1000):
    for number_b in range(100, 1000):
        product = number_a * number_b
        if str(product) == str(product)[::-1]:
            palindromes.append(product)

In [81]:
max(palindromes)

906609